// dotted-map/dist/without-countries.cjs
const require_chunk = require('./chunk-C0xms8kb.cjs');
let proj4 = require("proj4");
proj4 = require_chunk.__toESM(proj4);

//#region src/without-countries.ts
const PROJECTIONS = {
	mercator: "+proj=merc +lon_0={lng} +x_0=0 +y_0=0 +datum=WGS84 +units=m",
	equirectangular: "+proj=eqc +lon_0={lng} +lat_ts=0 +x_0=0 +y_0=0 +datum=WGS84 +units=m",
	robinson: "+proj=robin +lon_0={lng} +x_0=0 +y_0=0 +datum=WGS84 +units=m",
	equalEarth: "+proj=eqearth +lon_0={lng} +x_0=0 +y_0=0 +datum=WGS84 +units=m",
	mollweide: "+proj=moll +lon_0={lng} +x_0=0 +y_0=0 +datum=WGS84 +units=m",
	miller: "+proj=mill +lon_0={lng} +x_0=0 +y_0=0 +datum=WGS84 +units=m",
	sinusoidal: "+proj=sinu +lon_0={lng} +x_0=0 +y_0=0 +datum=WGS84 +units=m",
	orthographic: "+proj=ortho +lat_0={lat} +lon_0={lng} +x_0=0 +y_0=0 +datum=WGS84 +units=m",
	gallPeters: "+proj=cea +lon_0={lng} +lat_ts=45 +x_0=0 +y_0=0 +datum=WGS84 +units=m",
	vanDerGrinten: "+proj=vandg +lon_0={lng} +x_0=0 +y_0=0 +datum=WGS84 +units=m"
};
const DEFAULT_PROJECTION = { name: "mercator" };
const getProj4String = (projection = DEFAULT_PROJECTION) => {
	const template = PROJECTIONS[projection.name];
	const lat = projection.center?.lat ?? 0;
	const lng = projection.center?.lng ?? 0;
	return template.replace("{lat}", String(lat)).replace("{lng}", String(lng));
};
var DottedMapWithoutCountries = class {
	constructor({ map, avoidOuterPins = false }) {
		this.points = map.points;
		this.pins = [];
		this.X_MIN = map.X_MIN;
		this.Y_MAX = map.Y_MAX;
		this.X_RANGE = map.X_RANGE;
		this.Y_RANGE = map.Y_RANGE;
		this.grid = map.grid;
		this.width = map.width;
		this.height = map.height;
		this.ystep = map.ystep;
		this.avoidOuterPins = avoidOuterPins;
		this.proj4String = getProj4String(map.projection ?? DEFAULT_PROJECTION);
		this.image = {
			region: map.region,
			width: map.width,
			height: map.height
		};
	}
	addPin({ lat, lng, data, svgOptions }) {
		const pin = this.getPin({
			lat,
			lng
		});
		if (!pin) return void 0;
		const point = {
			...pin,
			data,
			svgOptions
		};
		this.pins.push(point);
		return point;
	}
	getPin({ lat, lng }) {
		const projected = (0, proj4.default)(this.proj4String, [lng, lat]);
		if (!projected.every((v) => Number.isFinite(v))) return;
		const [projX, projY] = projected;
		if (this.avoidOuterPins) return;
		let rawX = this.width * (projX - this.X_MIN) / this.X_RANGE;
		const rawY = this.height * (this.Y_MAX - projY) / this.Y_RANGE;
		const y = Math.round(rawY / this.ystep);
		if (y % 2 === 0 && this.grid === "diagonal") rawX -= .5;
		let localx = Math.round(rawX);
		const localy = Math.round(y) * this.ystep;
		if (y % 2 === 0 && this.grid === "diagonal") localx += .5;
		const inverse = (0, proj4.default)(this.proj4String, "WGS84", [localx * this.X_RANGE / this.width + this.X_MIN, this.Y_MAX - localy * this.Y_RANGE / this.height]);
		if (!inverse.every((v) => Number.isFinite(v))) return;
		const [localLng, localLat] = inverse;
		return {
			x: localx,
			y: localy,
			lat: localLat,
			lng: localLng
		};
	}
	getPoints() {
		return [...Object.values(this.points), ...this.pins];
	}
	getSVG({ shape = "circle", color = "current", backgroundColor = "transparent", radius = .5 } = {}) {
		const getPoint = ({ x, y, svgOptions = {} }) => {
			const pointRadius = svgOptions.radius || radius;
			if (shape === "circle") return `<circle cx="${x}" cy="${y}" r="${pointRadius}" fill="${svgOptions.color || color}" />`;
			else if (shape === "hexagon") {
				const sqrt3radius = Math.sqrt(3) * pointRadius;
				return `<polyline points="${[
					[x + sqrt3radius, y - pointRadius],
					[x + sqrt3radius, y + pointRadius],
					[x, y + 2 * pointRadius],
					[x - sqrt3radius, y + pointRadius],
					[x - sqrt3radius, y - pointRadius],
					[x, y - 2 * pointRadius]
				].map((point) => point.join(",")).join(" ")}" fill="${svgOptions.color || color}" />`;
			}
			return "";
		};
		return `<svg viewBox="0 0 ${this.width} ${this.height}" xmlns="http://www.w3.org/2000/svg" style="background-color: ${backgroundColor}">
        ${Object.values(this.points).map(getPoint).join("\n")}
        ${this.pins.map(getPoint).join("\n")}
      </svg>`;
	}
};

//#endregion
module.exports = DottedMapWithoutCountries;
